google/cloud/sql/connector/instance_connection_manager.py

```diff
@@ -406,66 +359,138 @@ async def seconds_until_refresh(self) -> int:
+                    expiration = token_expiration
+
+        except Exception as e:
+            logger.debug("Error occurred during _perform_refresh.")
+            raise e
+
         finally:
             self._refresh_in_progress.clear()
 
-        return refresh_task
+        return InstanceMetadata(
+            ephemeral_cert,
+            metadata["ip_addresses"],
+            priv_key,
+            metadata["server_ca_cert"],
+            expiration,
+            self._enable_iam_auth,
+        )
+
+    def _schedule_refresh(self, delay: int) -> asyncio.Task:
+        """
+        Schedule task to sleep and then perform refresh to get InstanceMetadata.
 
-    async def _schedule_refresh(self, delay: Optional[int] = None) -> asyncio.Task:
-        """A coroutine that sleeps for the specified amount of time before
-        running _perform_refresh.
+        :type delay: int
+        :param delay
+            Time in seconds to sleep before running _perform_refresh.
 
         :rtype: asyncio.Task
-        :returns: A Task representing _get_instance_data.
+        :returns: A Task representing the scheduled _perform_refresh.
         """
 
-        if delay is None:
-            delay = await self.seconds_until_refresh()
-        try:
-            logger.debug("Entering sleep")
-            await asyncio.sleep(delay)
-        except asyncio.CancelledError as e:
-            logger.debug("Schedule refresh task cancelled.")
-            raise e
-        return await self._perform_refresh()
+        async def _refresh_task(
+            self: InstanceConnectionManager, delay: int
+        ) -> InstanceMetadata:
+            """
+            A coroutine that sleeps for the specified amount of time before
+            running _perform_refresh.
+            """
+            refresh_task: asyncio.Task
+            try:
+                logger.debug("Entering sleep")
+                if delay > 0:
+                    await asyncio.sleep(delay)
+                refresh_task = self._loop.create_task(self._perform_refresh())
+                await refresh_task
+            except asyncio.CancelledError as e:
+                logger.debug("Schedule refresh task cancelled.")
+                raise e
+            # bad refresh attempt
+            except Exception as e:
+                logger.exception(
+                    "An error occurred while performing refresh. "
+                    "Scheduling another refresh attempt immediately",
+                    exc_info=e,
+                )
+                # check if current metadata is valid, don't want to replace valid metadata with invalid one
+                try:
+                    instance_data = await self._current
+                    if instance_data.expiration < datetime.datetime.now():
+                        raise ExpiredInstanceMetadata(
+                            "Instance metadata is expired, performing new refresh."
+                        )
```

Comment:
I feel like this bit is making the rest hard to read. What do you think about an `is_valid` function to help here:

```python
def is_valid(task: asyncio.Task[InstanceMetadata]) -> bool:
     try: 
          md = await task
         # only valid if now is before the cert expires
          if datetime.datetime.now() < instance_data.expiration:
              return True
      except Exception:
           # surpress any errors from task
      return False
```

Should help simplify this try/catch nest quite a bit